### PR TITLE
Make the state dict of CyclicLR scheduler pickleable

### LINKED ISSUE #15901: Error when checkpointing model with CyclicLR scheduler: `TypeError: cannot pickle 'WeakMethod' object`
### Bug description

In a module that uses a CyclicLR scheduler, training fails with the error `TypeError: cannot pickle 'WeakMethod' object`.

This is not caused by an issue with Lightning, but it took me quite a while to track down the issue, so perhaps there could be a check and/or nice warning message? Maybe the system should fall back to a simpler type of checkpoint (weights only) in cases where there's an error, or some un-picklable component?

### How to reproduce the bug

```python
1. Use a `torch.optim.lr_scheduler.CyclicLR` scheduler in a `pl.LightningModule`.
2. Train as usual (with `enable_checkpointing` defaulting to `True`)
3. After a bit Lightning will attempt to save a checkpoint, which will include the scheduler, but this scheduler includes a `WeakRef` (https://github.com/pytorch/pytorch/blob/4648baa91155718642937ec27b04196fde846566/torch/optim/lr_scheduler.py#L1226, in the PyTorch code) which can't be pickled.
```

### Error messages and logs

```
Traceback (most recent call last):
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3378, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-9-fa864ace7871>", line 209, in <module>
    trainer.fit(
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 582, in fit
    call._call_and_handle_interrupt(
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/call.py", line 38, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 624, in _fit_impl
    self._run(model, ckpt_path=self.ckpt_path)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1061, in _run
    results = self._run_stage()
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1140, in _run_stage
    self._run_train()
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1163, in _run_train
    self.fit_loop.run()
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/loops/loop.py", line 200, in run
    self.on_advance_end()
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/loops/fit_loop.py", line 295, in on_advance_end
    self.trainer._call_callback_hooks("on_train_epoch_end")
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1343, in _call_callback_hooks
    fn(self, self.lightning_module, *args, **kwargs)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 312, in on_train_epoch_end
    self._save_topk_checkpoint(trainer, monitor_candidates)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 371, in _save_topk_checkpoint
    self._save_none_monitor_checkpoint(trainer, monitor_candidates)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 660, in _save_none_monitor_checkpoint
    self._save_checkpoint(trainer, filepath)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/callbacks/model_checkpoint.py", line 374, in _save_checkpoint
    trainer.save_checkpoint(filepath, self.save_weights_only)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/trainer.py", line 1904, in save_checkpoint
    self._checkpoint_connector.save_checkpoint(filepath, weights_only=weights_only, storage_options=storage_options)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/trainer/connectors/checkpoint_connector.py", line 540, in save_checkpoint
    self.trainer.strategy.save_checkpoint(_checkpoint, filepath, storage_options=storage_options)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/pytorch_lightning/strategies/strategy.py", line 466, in save_checkpoint
    self.checkpoint_io.save_checkpoint(checkpoint, filepath, storage_options=storage_options)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/lightning_lite/plugins/io/torch_io.py", line 54, in save_checkpoint
    _atomic_save(checkpoint, path)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/lightning_lite/utilities/cloud_io.py", line 67, in _atomic_save
    torch.save(checkpoint, bytesbuffer)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/torch/serialization.py", line 423, in save
    _save(obj, opened_zipfile, pickle_module, pickle_protocol)
  File "/home/davidg/.virtualenvs/learning/lib/python3.10/site-packages/torch/serialization.py", line 635, in _save
    pickler.dump(obj)
TypeError: cannot pickle 'WeakMethod' object
```

### Environment

Version: `1.8.3.post1`

### More info

Explicitly passing `ModelCheckpoint(save_weights_only=True),` to the `Trainer.callbacks` parameter avoids the error.

cc @awaelchli

### LINKED ISSUE #90414: Cannot pickle 'WeakMethod' object when saving state_dict for CyclicLr
### 🐛 Describe the bug

When trying to save the CyclicLR state_dict and `scale_fn` is left at the default of `None` it will generate the error `cannot pickle 'WeakMethod' object`

```python
# All necessary imports at the beginning
import torch
from pathlib import Path

model = torch.nn.Sequential(
    torch.nn.Linear(3, 1),
    torch.nn.Flatten(0, 1)
)

kwargs = {'lr':             5.0e-6}
optimizer = torch.optim.AdamW(model.parameters(), **kwargs)

kwargs = {'base_lr':        5.0e-6,
          'max_lr':         2.5e-4,
          'cycle_momentum': False,
          'step_size_up':   100_000,
          'mode':           'exp_range',
          'gamma':          0.99999979673105275299293754990642}
scheduler = torch.optim.lr_scheduler.CyclicLR(optimizer, **kwargs)
scheduler_dict = scheduler.state_dict()

torch.save(scheduler_dict , Path('./checkpoint.pt'))
```
Outputs the following error:
```python
Traceback (most recent call last):
  File "E:\workspaces\CDVD-TSP\code\main_PWC-Net.py", line 1322, in <module>
    main()
  File "E:\workspaces\CDVD-TSP\code\main_PWC-Net.py", line 793, in main
    torch.save(model_state_dict, args.save_path / args.run_path / 'checkpoint.pt')
  File "C:\Users\Etienne\miniconda3\lib\site-packages\torch\serialization.py", line 423, in save
    _save(obj, opened_zipfile, pickle_module, pickle_protocol)
  File "C:\Users\Etienne\miniconda3\lib\site-packages\torch\serialization.py", line 635, in _save
    pickler.dump(obj)
TypeError: cannot pickle 'WeakMethod' object
```
Printing the dictionary showed 
```python
'_scale_fn_ref': <weakref at 0x000001E46B9E7C10; to 'CyclicLR' at 0x000001E47D9FE1F0>
```

Workaround by adding the following to delete the key
```python
scheduler_dict = scheduler.state_dict()

if '_scale_fn_ref' in scheduler_dict.keys():
    scheduler_dict.pop('_scale_fn_ref')

torch.save(scheduler_dict , Path('./checkpoint.pt'))
```

### Versions

```python
PyTorch version: 1.13.0
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Pro for Workstations
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.15 | packaged by conda-forge | (main, Nov 22 2022, 08:41:22) [MSC v.1929 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19044-SP0
Is CUDA available: True
CUDA runtime version: 11.7.99
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 526.86
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==1.13.0
[pip3] torch-lr-finder==0.2.1
[pip3] torchaudio==0.13.0
[pip3] torchvision==0.14.0
[conda] blas                      2.116                       mkl    conda-forge
[conda] blas-devel                3.9.0              16_win64_mkl    conda-forge
[conda] cudatoolkit               11.7.0              ha6f8bbd_10    conda-forge
[conda] libblas                   3.9.0              16_win64_mkl    conda-forge
[conda] libcblas                  3.9.0              16_win64_mkl    conda-forge
[conda] liblapack                 3.9.0              16_win64_mkl    conda-forge
[conda] liblapacke                3.9.0              16_win64_mkl    conda-forge
[conda] mkl                       2022.1.0           h6a75c08_874    conda-forge
[conda] mkl-devel                 2022.1.0           h57928b3_875    conda-forge
[conda] mkl-include               2022.1.0           h6a75c08_874    conda-forge
[conda] numpy                     1.23.5           py39hbccbffa_0    conda-forge
[conda] pytorch                   1.13.0          py3.9_cuda11.7_cudnn8_0    pytorch
[conda] pytorch-cuda              11.7                 h67b0de4_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch-lr-finder           0.2.1                    pypi_0    pypi
[conda] torchaudio                0.13.0                   pypi_0    pypi
[conda] torchvision               0.14.0                   pypi_0    pypi
```

cc @vincentqb @jbschlosser @albanD @janeyx99

### PR DESCRIPTION
**Pull request by awaelchli**
Fixes #90414

This PR drops the unpicklable `weakref.WeakMethod` object from CyclicLR scheduler from the state dict, and re-inits the object again once the state dict gets loaded. This makes the state picklable so you can include it in your checkpoint. Also fixes https://github.com/Lightning-AI/lightning/issues/15901

A simple test was added that `pickle.dumps(state)` the state.

### CODE REVIEW

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -1319,6 +1321,15 @@ def get_lr(self):
 
         return lrs
 
+    def state_dict(self):
+        # We are dropping the `_scale_fn_ref` attribute because it is a `weakref.WeakMethod` and can't be pickled
+        return {key: value for key, value in self.__dict__.items() if key not in ("optimizer", "_scale_fn_ref")}
+
+    def load_state_dict(self, state_dict):
+        self.__dict__.update(state_dict)
```
Comment (albanD) [inline torch/optim/lr_scheduler.py:1329]: `super().load_state_dict(state_dict)`

File: torch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -1319,6 +1321,15 @@ def get_lr(self):
 
         return lrs
 
+    def state_dict(self):
+        # We are dropping the `_scale_fn_ref` attribute because it is a `weakref.WeakMethod` and can't be pickled
+        return {key: value for key, value in self.__dict__.items() if key not in ("optimizer", "_scale_fn_ref")}
```
Comment (albanD) [inline torch/optim/lr_scheduler.py:1326]: You can do
res = super().state_dict()
res.pop("_scale_fn_ref")
return res

File: test/test_optim.py [hunk 3]
```diff
@@ -3158,6 +3160,35 @@ def test():
         assert ref() is None
         gc.enable()
 
+    def test_cycle_lr_state_dict_picklable(self):
+        adam_opt = optim.Adam(self.net.parameters())
+        scheduler = CyclicLR(adam_opt, base_lr=1, max_lr=5, cycle_momentum=False)
+        assert isinstance(scheduler._scale_fn_ref, weakref.WeakMethod)
```
Comment (albanD) [inline test/test_optim.py:3166]: self.assertIsInstance

File: test/test_optim.py [hunk 4]
```diff
@@ -3158,6 +3160,35 @@ def test():
...
+        state = scheduler.state_dict()
+        assert "_scale_fn_ref" not in state
```
Comment (albanD) [inline test/test_optim.py:3168]: self.assertNotIn

File: test/test_optim.py [hunk 5]
```diff
@@ -3158,6 +3160,35 @@ def test():
...
+        pickle.dumps(state)
+
+    def test_cycle_lr_scale_fn_restored_from_state_dict(self):
+        adam_opt = optim.Adam(self.net.parameters())
+
+        # Case 1: Built-in mode
+        scheduler = CyclicLR(adam_opt, base_lr=1, max_lr=5, cycle_momentum=False, mode="triangular2")
+        restored_scheduler = CyclicLR(adam_opt, base_lr=1, max_lr=5, cycle_momentum=False)
+        restored_scheduler.load_state_dict(scheduler.state_dict())
+        assert restored_scheduler.mode == scheduler.mode == "triangular2"
```
Comment (albanD) [inline test/test_optim.py:3178]: self.assertTrue

File: test/test_optim.py [hunk 6]
```diff
@@ -3158,6 +3160,35 @@ def test():
...
+        assert restored_scheduler._scale_fn_ref is not None and scheduler._scale_fn_ref is not None
```
Comment (albanD) [inline test/test_optim.py:3179]: self.assertIsNotNone

File: test/test_optim.py [hunk 7]
```diff
@@ -3158,6 +3160,35 @@ def test():
...
+
+        # Case 2: Custom `scale_fn`
+        def scale_fn(_):
+            return 0.5
+
+        scheduler = CyclicLR(adam_opt, base_lr=1, max_lr=5, cycle_momentum=False, scale_fn=scale_fn)
+        restored_scheduler = CyclicLR(adam_opt, base_lr=1, max_lr=5, cycle_momentum=False, scale_fn=scale_fn)
+        restored_scheduler.load_state_dict(scheduler.state_dict())
+        assert scheduler._scale_fn_custom is scale_fn
+        assert restored_scheduler._scale_fn_custom is scale_fn
```
Comment (albanD) [inline test/test_optim.py:3190]: self.assertIs here and above

**Review by albanD — COMMENTED** [review]
Thanks for the PR. It looks quite good, just small comments.

File: torch/optim/lr_scheduler.py (hunk 1 above)
Comment (awaelchli) [reply torch/optim/lr_scheduler.py]: Much better! Thanks for the comments, I addressed all of them.

**Review by albanD — APPROVED** [review]
update looks good, thanks

**Pull request closed without merging**